### PR TITLE
Adding close preview feature

### PR DESCRIPTION
**Pull request by americool**
**- Summary**

Added the basic functionality (via a button) to hide/show the preview panel in the Entry Editor.  It mirrors the request publish functionality. This was created to address issue #113. (*Note that it deals with the second level of integration @calavera mentioned in his most recent comment on the issue.*)

**- Test plan**

Tested code using the local "example" page. Added button(s) which call a function which toggles the state (*added to code as 'previewOpen'*). Added a showOnlyEditor, which displays the editor fullscreen without the preview. Depending on the status of this.state.previewOpen (*boolean*) the main render function displays either the original JSX or calls the showOnlyEditor. Attached gif below demonstrates its functionality: 

close_preview_pane

(*Note: Different styling maybe needed but css is not my strong-suit, and I was having some issues using the linked styles file. Would appreciate further direction or having someone with artistic ability step in on this part.*)

**- Description for the changelog**

Adds a Button allowing the preview pane to be toggled opened & closed. 

**- A picture of a cute animal (not mandatory but encouraged)**
*Same dog as last PR, different headgear.*
17191209_10101370520941889_309415902569031369_n

### CODE REVIEW

**Comment by erquhart** [comment]
@americool awesome contribution! 😎 We'll get this reviewed very soon.

**Comment by erquhart** [comment]
@americool we should be able to wrap the existing `PreviewPane` in a ternary, use a single button for toggling, and handle visibility and button label text based on state. Any reason not to go that route?

File: src/components/EntryEditor/EntryEditor.js [hunk 1]
```diff
@@ -39,9 +46,66 @@ class EntryEditor extends Component {
         onRemoveAsset,
         onCancelEdit,
     } = this.props;
+    const controlClassName = `${ styles.controlPane } ${ this.state.showEventBlocker && styles.blocker }`;
+    return (
```
Comment (Benaiah) [inline src/components/EntryEditor/EntryEditor.js]: I wouldn't copy all the markup when the only changes are wrapping with the SplitPane and removing the preview.

**Comment by americool** [comment]
I definitely wanted to conditionally render the preview pane originally, but I'm a little unclear on how to do it with the <SplitPane> component that's wrapping both the control pane and the preview pane. If we don't remove that component as well we simply end up rendering a blank pane next to the editor rather than have it go full screen. Any thoughts?

**Comment by bdougie** [comment]
@erquhart I tried this locally with suggestion and confirmed that only hides the `<PreviewPane />` and does not expand the editor. This is because the use of fixed width. I chatted with @americool and saw that conditionally rendering the defaultSize of the SplitPane, actually breaks dragging as well.

ezgif com-video-to-gif 4

What are your thoughts on moving the `showOnlyEditor` and the content of the EntryEditor Return to their own component and render conditionally.

```js
return this.state.previewOpen ? <ShowSplitPaneEditor {...this.props} /> : <ShowOnlyEditor {...this.props} />
``` 

If the above is not ideal, the other solution could be to rethink the CSS, maybe with flexbox (removing the fixed widths) which will adjust the width automatically.

**Comment by erquhart** [comment]
@bdougie agreed, rendering conditionally makes sense, and I don't even think separate components are necessary, the swappable template portions can just be saved to variables right in the render function.

@americool sound good?

**Comment by americool** [comment]
So I'd just finished writing a version with two components before reading this. Feel free to look into that code. I'll try to attempt a conditional render with save variables now. May have some questions along the way. Thanks for the feedback!

**Comment by erquhart** [comment]
A cheap improvement that should probably be present at merge time is persisting the preview visibility state. We can follow the method used for persisting the rich text editor visual mode: https://github.com/netlify/netlify-cms/blob/a4d7622adead7e259caccdac5c97499d977de070/src/components/Widgets/MarkdownControl.js#L23

**Comment by americool** [comment]
I may not understand how to appropriately save the swappable template portions into variables, because as I start to code it, looking at two separate consts that hold all the relative jsx, which are both already very long, will also *each* have to have all the destructed props passed in, along with the styles, classnames, and functions. It seems *really* long and ugly so I figure I must be approaching this wrong, because as it stands the separate files at least seem to look a lot more elegant. Happy to do it either way, but I want to make sure what I'm doing makes sense before I go down a foolish path.

**Comment by erquhart** [comment]
@americool like this:

```javascript
  render() {
    const {
        collection,
        entry,
        fields,
        fieldsMetaData,
        fieldsErrors,
        getAsset,
        onChange,
        onValidate,
        onAddAsset,
        onRemoveAsset,
        onCancelEdit,
    } = this.props;

    const controlClassName = `${ styles.controlPane } ${ this.state.showEventBlocker && styles.blocker }`;
    const previewClassName = `${ styles.previewPane } ${ this.state.showEventBlocker && styles.blocker }`;
    const editor = (
      <div className={controlClassName}>
        <ControlPane
          collection={collection}
          entry={entry}
          fields={fields}
          fieldsMetaData={fieldsMetaData}
          fieldsErrors={fieldsErrors}
          getAsset={getAsset}
          onChange={onChange}
          onValidate={onValidate}
          onAddAsset={onAddAsset}
          onRemoveAsset={onRemoveAsset}
          ref={c => this.controlPaneRef = c} // eslint-disable-line
        />
      </div>
    );
    const editorWithPreview = (
      <ScrollSync>
        <div className={styles.container}>
          <SplitPane
            defaultSize="50%"
            onDragStarted={this.handleSplitPaneDragStart}
            onDragFinished={this.handleSplitPaneDragFinished}
          >
            <ScrollSyncPane>
              {editor}
            </ScrollSyncPane>
            <div className={previewClassName}>
              <PreviewPane
                collection={collection}
                entry={entry}
                fields={fields}
                fieldsMetaData={fieldsMetaData}
                getAsset={getAsset}
              />
            </div>
          </SplitPane>
        </div>
      </ScrollSync>
    );
    return (
      <div className={styles.root}>
        {this.state.showPreview ? editor : editorWithPreview}
        <div className={styles.footer}>
          <Toolbar
            isPersisting={entry.get('isPersisting')}
            onPersist={this.handleOnPersist}
            onCancelEdit={onCancelEdit}
          />
        </div>
      </div>
    );
  }
```

**Comment by americool** [comment]
Thanks, sorry about the confusion. I see the mistake I was making now, I wasn't creating the constants correctly which messed up my scope so I couldn't access all the other data easily.  (*it also didn't occur to me to toss the editor component inside the editorwithpreview*) Thanks for the help! I'll have something up soon!

**Comment by americool** [comment]
I think at this point it's ready for further review.

File: src/components/EntryEditor/EntryEditor.js [hunk 2]
```diff
@@ -1,15 +1,19 @@
...
 import PreviewPane from '../PreviewPane/PreviewPane';
 import Toolbar from './EntryEditorToolbar';
 import styles from './EntryEditor.css';
 
+const PREVIEW_STATE = 'cms.preview-state';
+
 class EntryEditor extends Component {
   state = {
     showEventBlocker: false,
+    previewOpen: localStorage.getItem(PREVIEW_STATE) === "open",
```
Comment (erquhart) [inline src/components/EntryEditor/EntryEditor.js]: We should simplify this by using a boolean - name it `PREVIEW_HIDE` and set to `'true'` on hide. It'll still be stored as a string, but it's better than checking for custom string values.

File: src/components/EntryEditor/EntryEditor.js [hunk 3]
```diff
@@ -40,56 +53,63 @@ class EntryEditor extends Component {
...
-                    getAsset={getAsset}
-                    onChange={onChange}
-                    onValidate={onValidate}
-                    onAddAsset={onAddAsset}
-                    onRemoveAsset={onRemoveAsset}
-                    ref={c => this.controlPaneRef = c} // eslint-disable-line
-                  />
+    const editor = (
+      <div className={controlClassName}>
+        <Button onClick={this.handleTogglePreview}> TOGGLE PREVIEW </Button>
```
Comment (erquhart) [inline src/components/EntryEditor/EntryEditor.js]: No spacing around the button label text.

File: src/components/EntryEditor/EntryEditor.js [hunk 4]
```diff
@@ -40,56 +53,63 @@ class EntryEditor extends Component {
...
-                  entry={entry}
-                  fields={fields}
-                  fieldsMetaData={fieldsMetaData}
-                  getAsset={getAsset}
-                />
-              </div>
-            </SplitPane>
-          </div>
-        </ ScrollSync>
+    const editorWithPreivew = (
```
Comment (erquhart) [inline src/components/EntryEditor/EntryEditor.js]: Fix spelling - `editorWithPreview`.

**Comment by americool** [comment]
Thanks for the feedback. The most recent commit should take care of those issues.

**Pull request merged**